Makefile

```diff
@@ -1,9 +1,359 @@
+
+# Uncomment this to see symbol, for debugging.
+# $(info $$TEST_C_CODE_SRC     = [${TEST_C_CODE_SRC} ])
+# $(info $$TEST_CPP_CODE_SRC   = [${TEST_CPP_CODE_SRC} ])
+
+# ----------------------------------------------------------------------------
+# Generate location / name of test-use-cases sample program binary, using
+# built-in substitution references. test-use-cases/ has sources like:
+#
+#   test-use-cases/single-file-C-program/test-main.c
+#   test-use-cases/single-file-Cpp-program/test-main.cpp
+#
+# Convert this list of *main.c to generate test-code program binary names,
+# using the src-dir's name as the program name.
+#
+# Generate, respectively:
+#
+#   build/release/bin/test-use-cases/single-file-C-program
+#   build/release/bin/test-use-cases/single-file-Cpp-program
+#
+TEST_C_CODE_BINS_TMP := $(TEST_C_CODE_SRC:$(TESTSDIR)/%-main.c=$(BINDIR)/$(TESTSDIR)/%-program)
+TEST_C_CODE_BINS=$(patsubst %program/, %program, $(dir $(TEST_C_CODE_BINS_TMP)) )
+
+# Uncomment this to see symbol, for debugging.
+# $(info $$TEST_C_CODE_BINS     = [${TEST_C_CODE_BINS} ])
+
+TEST_CPP_CODE_BINS_TMP := $(TEST_CPP_CODE_SRC:$(TESTSDIR)/%-main.cpp=$(BINDIR)/$(TESTSDIR)/%-program)
+TEST_CPP_CODE_BINS_TMP := $(TEST_CPP_CODE_BINS_TMP:$(TESTSDIR)/%-main.cc=$(BINDIR)/$(TESTSDIR)/%-program)
+TEST_CPP_CODE_BINS=$(patsubst %program/, %program, $(dir $(TEST_CPP_CODE_BINS_TMP)) )
+
+TEST_CC_CODE_BINS_TMP := $(TEST_CPP_CODE_SRC:$(TESTSDIR)/%-main.cpp=$(BINDIR)/$(TESTSDIR)/%-program)
+TEST_CC_CODE_BINS_TMP := $(TEST_CPP_CODE_BINS_TMP:$(TESTSDIR)/%-main.cc=$(BINDIR)/$(TESTSDIR)/%-program)
+TEST_CC_CODE_BINS=$(patsubst %program/, %program, $(dir $(TEST_CPP_CODE_BINS_TMP)) )
+
+# Uncomment this to see symbol, for debugging.
+# $(info $$TEST_CPP_CODE_BINS     = [${TEST_CPP_CODE_BINS} ])
+
+# Build symbol for single C & C++ unit-test binary that we run
+C_UNIT_TEST_BIN     := $(BUILD_ROOT)/$(BUILD_MODE)/bin/test-use-cases/single-file-C-program
+CPP_UNIT_TEST_BIN   := $(BUILD_ROOT)/$(BUILD_MODE)/bin/test-use-cases/single-file-Cpp-program
+CC_UNIT_TEST_BIN    := $(BUILD_ROOT)/$(BUILD_MODE)/bin/test-use-cases/single-file-CC-program
+
+# ###################################################################
+# Report build machine details and compiler version for troubleshooting,
+# so we see this output for clean builds, especially in CI-jobs.
+# ###################################################################
+.PHONY : clean tags
+clean:
+	uname -a
+	$(CC) --version
+	rm -rf $(BUILD_ROOT)
+
+# Delete l3.o object as it needs to be recompiled for inclusion w/C++ sources
+clean-l3:
+	rm -rf $(OBJDIR)/src/l3.o
+
+# ###################################################################
+# Make build targets begin here
+# ###################################################################
+
+all-c-tests:   $(TEST_C_CODE_BINS)
+all-cpp-tests: $(TEST_CPP_CODE_BINS)
+all-cc-tests:  $(TEST_CC_CODE_BINS)
+
```

Comment:
Note, unlike normal Makefiles, this can no longer supply a single `all` target which will subsume all these `all*` flavours.

This restriction is because of the inability to mix objects built with gcc and g++ in one link step (which I explained at the top).

As this Makefile is mainly for a how-to demo exercise, I think it's ok to live with these multiple `all*` targets.